src/controllers/orderController.js

```diff
@@ -5,58 +5,103 @@ const { BadRequestError } = require('../errors');
 const PurchasedAlbum = require('../models/PurchasedAlbum');
 const CustomError = require('../errors');
 const { checkPermissions } = require('../utils');
+const { sendOrderCompletedEmail } = require('../mailing/sender');
+const mongoose = require('mongoose');
 
 const createOrder = async (req, res) => {
-  const { orderItems, subtotal, tax, total } = req.body;
+  const session = await mongoose.startSession();
+
+  try {
+    session.startTransaction();
+
+    const { orderItems, subtotal, tax, total } = req.body;
+
+    if (!orderItems?.length) {
+      throw new BadRequestError(
+        'Unable to create an order because the order items are missing.'
+      );
+    }
+    // Fetch user information from the request object
+    const user = req.user;
 
-  // Perform necessary checks and validations
-  if (!orderItems?.length) {
-    throw new BadRequestError(
-      'Unable to create order because the order items are missing.'
-    );
-  }
 
-  // Create an Order document in the database
-  const order = await Order.create({
-    user: req.user.userId,
-    orderItems,
-    subtotal,
-    tax,
-    total,
-  });
 
-  // Create a payment intent with Stripe
-  const paymentIntent = await stripe.paymentIntents.create({
-    amount: total * 100, // Stripe requires amount in cents
-    currency: 'usd',
-    description: `Order #${order._id}`,
-    metadata: {
-      userId: req.user.userId,
-      orderId: order._id, // key-values
-      totalQuantity: orderItems.reduce(
-        (total, item) => total + item.quantity,
-        0
-      ), // the total number of items in the order
-      totalAlbums: orderItems.length, // the total number of albums purchased in order
-    },
-  });
 
-  // Save the paymentIntent to the Order model
-  order.paymentIntentId = paymentIntent.id;
-  await order.save();
 
-  // Create PurchasedAlbum entries for each album in orderItems
-  for (const orderItem of orderItems) {
-    await PurchasedAlbum.create({
-      album: orderItem.album, // album ID from orderItem
-      user: req.user.userId, // user ID from the request
+// Populate the orderItems with full album data
+const populatedOrderItems = await Promise.all(
+  orderItems.map(async (orderItem) => {
+    return await orderItem.populate('album').execPopulate();
+  })
+);
+  // Create an array to store the order items with full album data
+  const orderItemsWithFullAlbum = populatedOrderItems.map((orderItem) => ({
+    album: orderItem.album.toObject(), // Use toObject() to convert the full album to a plain JavaScript object
+    quantity: orderItem.quantity,
+  }));
```

Comment:
Could you explain why we are trying to populate the album object in order to save it here? Saving the album id should be sufficient and if we need the full object then we can populate it when _fetching_ from the database instead.

Reply:
@akosasante I was trying to get the whole object ... now added populate method to Order.js to .post('save') middleware 
```

// Middleware to send order completion email when the order status changes to "complete"
OrderSchema.post('save', async function (doc) {
  if (doc.orderStatus === 'complete') {
    const orderItemsWithFullAlbum = await doc
      .populate({ path: 'orderItems.album' })
      .execPopulate();

    // Send the order completion email
    await sendOrderCompletedEmail(
      doc.user.email,
      doc.user.name,
      orderItemsWithFullAlbum.orderItems,
      orderItemsWithFullAlbum.total
    );
  }
});
```